### PR TITLE
Enable toggling of video stretching

### PR DESCRIPTION
**Pull request by rschroll**
This adds a pane that slides out from the right side of the window to adjust settings.  It's triggered by a little handle that overlays the top right corner of the screen.  This seems to me to be the best for right-handers, since it's often behind their hand, but is unlikely to be hit by their wrist.  Perhaps there should be a handedness option to let it display on the left as well.

Settings are persisted in localstorage, since the server doesn't need to know about this setting.  But if future settings impact the server, perhaps they should be cookies, so their automatically delivered there.

When I load the page on my tablet, sometimes it zooms too wide, to show the off-screen panel as well.  Toggling it fixes it.  There's probably some meta property to tell it not to do this zooming, but I haven't investigated yet.

### CODE REVIEW

**Comment by rschroll** [comment]
I was going to wait for feedback on the UI idea before continuing, but I was curious about the changes in the html_config branch.  And I ended up building out the rest of the UI.  Let me know what you think.

**Comment by H-M-H** [comment]
Overall I like it :)

> It's triggered by a little handle that overlays the top right corner of the screen. 

I guess this is the best option we have. I would like it to disappear completely though but that's probably impossible.

> Perhaps there should be a handedness option to let it display on the left as well.

Definitely, something else I want to avoid is to make some part of the touch area inaccessible (so if you really want to it should be possible to send inputs to the top right corner). Some option to delete the menu for one session (until next page reload) is probably a good idea too.

> localstorage

You are right, I did not think about that, localstorage is the better choice here.

---
Some more thoughts:
- Ignoring touch, mouse or stylus events should probably be done in the browser now.
- ConfigError and ConfigOk need to be handled

Finally I would like to ask you to add yourself to https://github.com/H-M-H/Weylus/blob/master/CONTRIBUTORS and by doing so license your contributions under the 3-Clause BSD License.

**Comment by rschroll** [comment]
> I would like it to disappear completely though but that's probably impossible.

I initially had a clever idea to hide the UI completely when the video had focus.  That way, it would go away when you started using it as a stylus, but it would return when you click on some browser chrome.  This worked with a mouse, but apparently touch events don't cause focus to be grabbed.  So on a touch device, the UI never disappeared.

Another approach would be to have some gesture that would indicate Weylus should grab back control and display the UI.  But this would have to be complicated enough to not conflict with any other software.  This would then cause trouble as we try to figure out what to do with events that may or may not be a part of that control gesture.

On somewhat promising gesture would be a drag from off the edge of the screen, which would be unlikely to be a stylus action.  But this will conflict with some tablet OS gestures.  (Android has recently made that mean "back".)  So I don't think this works, either.

> Definitely, something else I want to avoid is to make some part of the touch area inaccessible (so if you really want to it should be possible to send inputs to the top right corner). Some option to delete the menu for one session (until next page reload) is probably a good idea too.

Both should be possible without too much hassle.  One good result of turning off video stretching is that it's highly unlikely that there is content in the corners.  Any slight mismatch in aspect ratio will result in letterboxing or whatever the orthogonal version of that is called.

> Ignoring touch, mouse or stylus events should probably be done in the browser now.

It looks like, right now, the browser handler doesn't distinguish between the event types.  Is that right?  Adding the check shouldn't be too hard.  Will these settings still need to be sent back to the server?  It looks like they're being used to set up the different input methods.

> ConfigError and ConfigOk need to be handled

I was thinking of a full-page overlay to display disconnection events.  Perhaps that could be used for ConfigError, too.  What needs to be done for ConfigOk?

> Finally I would like to ask you to add yourself to https://github.com/H-M-H/Weylus/blob/master/CONTRIBUTORS and by doing so license your contributions under the 3-Clause BSD License.

Will do!

**Comment by H-M-H** [comment]
>  hide the UI completely when the video had focus

Some other disadvantage is that it probably does not work in full screen mode (at least on iOS/iPadOS you can add a bookmark to your home screen that lets you open the bookmarked page in full screen mode).

> On somewhat promising gesture would be a drag from off the edge of the screen, which would be unlikely to be a stylus action. But this will conflict with some tablet OS gestures.

Swiping from the right edge should probably work but I am not sure if I like this. Would need some testing if it really does not interfere with normal usage.

Some other idea: What do you think about hiding the menu once the video element receives pointer events and only making it reappear after there have not been any pointer events for say 5 seconds?

> It looks like, right now, the browser handler doesn't distinguish between the event types. Is that right?

That is right, the Pointerhandler just sends every event received and needs to be patched.

> Adding the check shouldn't be too hard. Will these settings still need to be sent back to the server? It looks like they're being used to set up the different input methods.

Previously that was the case, I removed these settings from the rust side with 993a72e727975ce3a81ca0633518bafc7c74b9e8 and now they should be browser/client only.

> I was thinking of a full-page overlay to display disconnection events. Perhaps that could be used for ConfigError, too.

I think a ConfigError should be handled differently as it only needs to be displayed to the user whereas a disconnection event should offer reconnecting (which is done through a page reload right now).

>  What needs to be done for ConfigOk?

ConfigOk should be used to kick off video streaming by sending the first "GetFrame" if the configuration succeeded.

Currently the protocol is supposed to work like this:
```
Client: GetCapturableList
Server: CapturableList
Client: Config
Server: ConfigOk or ConfigError
Client: if ConfigOk: GetFrame and PointerEvent else display error
Server: Error if something went wrong
```

Something I still need decide is how to handle missing options as on Windows and macOS sending the Config does not actually configure anything.

**Comment by rschroll** [comment]
These commits add client-side filtering of events by pointer type.  However, they still report the new configuration back to the server, and they rely on the server's reply to trigger the change in processing.  If the server doesn't need to know about these changes, this bit should be reworked.  The UI panel can now be swapped left and right and "vanished" completely.

Other issues:
> Swiping from the right edge should probably work but I am not sure if I like this. Would need some testing if it really does not interfere with normal usage.

Android uses this as a back action, so it won't work everywhere.

> Some other idea: What do you think about hiding the menu once the video element receives pointer events and only making it reappear after there have not been any pointer events for say 5 seconds?

That's an interesting idea.  I think it only fails if you can only touch the area under the handle -- there's no way to make the handle go away to allow this.  That said, the vanish option I put in works well enough, I think, and is dead simple.  So I'd probably argue against this.

> ConfigOk should be used to kick off video streaming by sending the first "GetFrame" if the configuration succeeded.

This was already set up in the code you implemented :).  I think the only difference between your protocol description and the current state of the branch is that the client doesn't wait for the capturable list.  Instead, it just asks for a stream of capturable #0, which seems to always be the desktop.  Then when the list arrives, the user can change this.  I sort of like this approach, since it starts working immediately, rather than waiting for user input.

**Comment by H-M-H** [comment]
> However, they still report the new configuration back to the server, and they rely on the server's reply to trigger the change in processing.

This should indeed be changed, right now everything the server needs to know is https://github.com/H-M-H/Weylus/blob/21a3a10b3af93b9e1fdbc329bdc522133ff363e8/src/protocol.rs#L5-L8 and event filtering should not concern the server.

>  The UI panel can now be swapped left and right and "vanished" completely.

Awesome!

> Android uses this as a back action, so it won't work everywhere.

I did not realize that, so this is basically the opposite of what iPadOS does because here swiping from the left edge to the right is the "back" action...

> That said, the vanish option I put in works well enough, I think, and is dead simple. So I'd probably argue against this.

I think I am fine with vanishing the menu(handle), no need to over complicate things if this works well enough.

> Instead, it just asks for a stream of capturable #0, which seems to always be the desktop. Then when the list arrives, the user can change this. I sort of like this approach, since it starts working immediately, rather than waiting for user input.

Agreed, this does seem a bit nicer than what I originally intended (waiting for user input) and capturable #0 indeed is always the desktop.

I think there are only two things missing now:
- Refreshing the capturables list
- setting a delay between "GetFrame" requests

File: www/static/style.css [hunk 1]
```diff
@@ -22,3 +22,91 @@ video {
...
+    left: -1em;
+    width: 1em;
+    border-radius: 0 0 0 0.25em;
+    padding: 0.2em 0 0.1em 0;
+    background: #111c;
+    color: #444;
+    font-size: 150%;
+    text-align:center;
+}
+.lefty #handle {
```
Comment (H-M-H) [inline www/static/style.css:60]: Looks like `border-radius` and `padding` should be set here too.

**Comment by H-M-H** [comment]
One more thing: you might want to rebase against the latest version as this fixes input getting progressively slower.

**Comment by H-M-H** [comment]
Alright, I will merge this and continue work from there. Thanks for your contribution!

**Comment by rschroll** [comment]
Thanks for getting it all merged in.  Sorry I went dark -- I suddenly got very busy with work.

**Pull request merged**